File: app/src/main/java/com/arejas/dashboardofthings/data/sources/network/NetworkInterfaceHelper.java
package com.arejas.dashboardofthings.data.sources.network;

import android.content.Context;

import com.arejas.dashboardofthings.domain.entities.database.Actuator;
import com.arejas.dashboardofthings.domain.entities.database.Network;
import com.arejas.dashboardofthings.domain.entities.database.Sensor;

import java.util.HashMap;
import java.util.Map;

public abstract class NetworkInterfaceHelper {

    private Network network;

    private Map<Integer, Sensor> sensorsRegistered;

    private final Object sensorsLock = new Object();

    public NetworkInterfaceHelper(Network network) {
        this.network = network;
        sensorsRegistered = new HashMap<>();
    }

    public abstract boolean initNetworkInterface(Context context, Sensor[] sensors);

    public abstract boolean closeNetworkInterface(Context context);

    public abstract boolean configureSensorReceiving(Context context, Sensor sensor);

    public abstract boolean unconfigureSensorReceiving(Context context, Sensor sensor);

    public abstract boolean sendActuatorData(Context context, Actuator actuator, String dataToSend);

    public abstract boolean requestSensorReload(Context context, Sensor sensor);

    public Network getNetwork() {
        return network;
    }

    public void setNetwork(Network network) {
        this.network = network;
    }

    public Map<Integer, Sensor> getSensorsRegistered() {
        synchronized (sensorsLock) {
            Map<Integer, Sensor> returnedMap = new HashMap<>();
            returnedMap.putAll(sensorsRegistered);
            return returnedMap;
        }
    }

    public void setSensorsRegistered(Map<Integer, Sensor> sensorsRegistered) {
        synchronized (sensorsLock) {
            this.sensorsRegistered = sensorsRegistered;
        }
    }

    public void registerSensor(Sensor sensor) {
        synchronized (sensorsLock) {
            this.sensorsRegistered.put(sensor.getId(), sensor);
        }
    }

    public void unregisterSensor(Sensor sensor) {
        synchronized (sensorsLock) {
            this.sensorsRegistered.remove(sensor.getId());
        }
    }
}

File: app/src/main/java/com/arejas/dashboardofthings/data/sources/network/NetworkInterfaceHelper.kt
package com.arejas.dashboardofthings.data.sources.network

import android.content.Context

import com.arejas.dashboardofthings.domain.entities.database.Actuator
import com.arejas.dashboardofthings.domain.entities.database.Network
import com.arejas.dashboardofthings.domain.entities.database.Sensor

import java.util.HashMap

abstract class NetworkInterfaceHelper(var network: Network) {

    internal var sensorsRegistered: MutableMap<Int, Sensor>

    private val sensorsLock = Any()

    init {
        sensorsRegistered = HashMap()
    }

    abstract fun initNetworkInterface(context: Context, sensors: Array<Sensor>): Boolean

    abstract fun closeNetworkInterface(context: Context): Boolean

    abstract fun configureSensorReceiving(context: Context, sensor: Sensor): Boolean

    abstract fun unconfigureSensorReceiving(context: Context, sensor: Sensor): Boolean

    abstract fun sendActuatorData(context: Context, actuator: Actuator, dataToSend: String): Boolean

    abstract fun requestSensorReload(context: Context, sensor: Sensor): Boolean

    fun getSensorsRegistered(): Map<Int, Sensor> {
        synchronized(sensorsLock) {
            val returnedMap = HashMap<Int, Sensor>()
            returnedMap.putAll(sensorsRegistered!!)
            return returnedMap
        }
    }

    fun setSensorsRegistered(sensorsRegistered: MutableMap<Int, Sensor>) {
        synchronized(sensorsLock) {
            this.sensorsRegistered = sensorsRegistered
        }
    }

    fun registerSensor(sensor: Sensor) {
        synchronized(sensorsLock) {
            this.sensorsRegistered!!.put(sensor.id!!, sensor)
        }
    }

    fun unregisterSensor(sensor: Sensor) {
        synchronized(sensorsLock) {
            this.sensorsRegistered!!.remove(sensor.id)
        }
    }
}
